src/backend/distributed/planner/multi_master_planner.c

```diff
@@ -285,6 +364,81 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
 }
 
 
+static PlannedStmt *
+BuildSelectStatementViaStdPlanner(Query *masterQuery, List *masterTargetList,
+								  CustomScan *remoteScan)
+{
+	PlannerGlobal *glob = makeNode(PlannerGlobal);
+	PlannerInfo *root = makeNode(PlannerInfo);
+	root->parse = masterQuery;
+	root->glob = glob;
+	root->query_level = 1;
+	root->planner_cxt = CurrentMemoryContext;
+	root->wt_param_id = -1;
+
+	/* reset state */
+	glob = makeNode(PlannerGlobal);
+	root = makeNode(PlannerInfo);
+	root->parse = masterQuery;
+	root->glob = glob;
+	root->query_level = 1;
+	root->planner_cxt = CurrentMemoryContext;
+	root->wt_param_id = -1;
+
+	/*
+	 * the standard planner will scribble on the target list, since we need the target
+	 * list for alias creation we make a copy here.
+	 * TODO might be able to prevent the copy if we can move the alias creation before the standard planner
+	 */
+
+	remoteScan->custom_scan_tlist = copyObject(masterTargetList);
+	remoteScan->scan.plan.targetlist = copyObject(masterTargetList);
+
+	/* probably want to do this where we add sublinks to the master plan */
+	masterQuery->hasSubLinks = checkExprHasSubLink((Node *) masterQuery);
+
+	/* This code should not be re-entrant */
+	PlannedStmt *standardStmt = NULL;
+	PG_TRY();
+	{
+		Assert(ReplaceCitusExtraDataContainer == false);
+		Assert(ReplaceCitusExtraDataContainerWithCustomScan == NULL);
+		ReplaceCitusExtraDataContainer = true;
+		ReplaceCitusExtraDataContainerWithCustomScan = remoteScan;
+
+		standardStmt = standard_planner(masterQuery, 0, NULL);
+
+		ReplaceCitusExtraDataContainer = false;
+		ReplaceCitusExtraDataContainerWithCustomScan = NULL;
+	}
+	PG_CATCH();
+	{
+		ReplaceCitusExtraDataContainer = false;
+		ReplaceCitusExtraDataContainerWithCustomScan = NULL;
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	Assert(standardStmt != NULL);
+
+	List *columnNameList = NIL;
+	TargetEntry *targetEntry = NULL;
+
+	/*
+	 * (7) Replace rangetable with one with nice names to show in EXPLAIN plans
+	 */
+	foreach_ptr(targetEntry, masterTargetList)
+	{
+		columnNameList = lappend(columnNameList, makeString(targetEntry->resname));
+	}
+
+	RangeTblEntry *customScanRangeTableEntry = linitial(standardStmt->rtable);
```

Comment:
The assumption that the first item in the rtable is always the custom scan RTE seems a little fragile. Any other way to do this?

Reply:
You are completely correct! This code came from the earlier implementation.

Ideally we would have the range table entries from the original query and label the fields coming from the `CustomScan` appropriately instead of relabelling the fields to come from `remote_scan`.

For now I think we can look for the first range table entry that refers to the citus_extra_datacontainer function as that is the range table entry we should be changing the alias for.